regression_report.py

```diff
@@ -0,0 +1,143 @@
+            exe = reports["exe"]
+            rev = reports["rev"]
+            # remember only the first (latest) revision for the given executable
+            if exe not in revisions:
+                revisions[exe] = rev
+    return revisions
+
+"""
+Returns a dict executable -> benchmark names 
+"""
+def loadBenchmarkNames(codespeedUrl):
+    execToBenchmarks = {}
+    revisions = loadExecutableAndRevisions(codespeedUrl)
+    # Per each revision/executable ask for benchmark names
+    # http://codespeed.dak8s.net:8000/changes/table/?tre=10&rev=b9593715f35-ffcdbb4-1647399268&exe=1&env=2
+    for exe, rev in revisions.items():
+        url = codespeedUrl + 'changes/table/?' + urllib.urlencode({'tre': '10', 'rev': rev, 'exe': exe, 'env': ENVIRONMENT})
+        f = urllib2.urlopen(url)
+        response = f.read()
+        f.close()
+        benchmarks = []
+        for line in response.split("\n"):
+            # look for lines like and extract benchmark name between characters > and <
+            # <td title="">remoteSortPartition</td>
+            if '<td title="">' in line:
+                benchmark = re.findall(r'\>([^<]+)\<', line)[0]
+                benchmarks.append(benchmark)
+        execToBenchmarks[exe] = benchmarks
+    return execToBenchmarks
+
+"""
+Returns a list of benchmark results
+"""
+def loadData(codespeedUrl, exe, benchmark, downloadSamples):
+    url = codespeedUrl + 'timeline/json/?' + urllib.urlencode({'exe': exe, 'ben': benchmark, 'env': ENVIRONMENT, 'revs': downloadSamples})
+    f = urllib2.urlopen(url)
+    response = f.read()
+    f.close()
+    result = json.loads(response)['timelines'][0]['branches']['master'][exe]
+    return [score for (date, score, deviation, commit, branch) in result]
+
+def getMedian(lst):
+    lst = sorted(lst)  # Sort the list first
+    if len(lst) % 2 == 0:  # Checking if the length is even
+        # Applying formula which is sum of middle two divided by 2
+        return (lst[len(lst) // 2] + lst[(len(lst) - 1) // 2]) / 2
+    else:
+        # If length is odd then get middle value
+        return lst[len(lst) // 2]
+
+def checkBenchmark(args, exe, benchmark):
+    results = loadData(args.codespeed, exe, benchmark, args.downloadSamples)
+
+    urlToBenchmark = args.codespeed + 'timeline/#/?' + urllib.urlencode({
+        'ben': benchmark,
+        'exe': exe,
+        'env': ENVIRONMENT,
+        'revs': args.downloadSamples,
+        'equid': 'off',
+        'quarts': 'on',
+        'extr': 'on'})
+
+    if len(results) < args.baseLine:
+        return
+    median = getMedian(results[args.recentTrend:args.baseLine])
+    recent_median = getMedian(results[:args.recentTrend])
+    median_trend = 0
+    if median != 0:
+        median_trend = recent_median * 100 / median - 100
+    if median_trend < args.medianTrendThreshold:
+        # regression detected, print warning
+        print "<%s|%s> median=%s recent_median=%s" % (urlToBenchmark, benchmark, median, recent_median)
```

Comment:
How or where should this script work? I mean I thought that we have jenkins job which periodically can check for regression and publish the result to the slack. But here I see only printing into the console. So what the main scenario of usage of this script?

Reply:
As described offline. Jenkins is reading stdout from this script and pushing it to our monitoring system.